### PR TITLE
Add recipe for logms

### PR DESCRIPTION
**Pull request by jcs090218**
### Brief summary of what the package does

See where the message came from

### Direct link to the package repository

https://github.com/jcs-elpa/logms

### Your association with the package

The maintainer

### Relevant communications with the upstream package maintainer

None needed

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### CODE REVIEW

**Comment by jcs090218** [comment]
Sorry for the last pull request #7628. I figured out there was a huge bug but now it's fix!

**Comment by alphapapa** [comment]
IIUC this simply adds a prefix to `message` messages.  Couldn't that be done with a simple advice function?

**Comment by jcs090218** [comment]
Can you achieve this by using advice? I don't want to pollute message function so I think it's better to use a new function call.

**Comment by riscy** [comment]
I see.  I think the package summary "see where the message came from" creates some of the confusion and should probably be revised.  I'd say this package provides a logging function whose output contains clickable links to context. 

The code took me a while to go through.  This strikes me as a very cool idea, but I think the implementation is a bit rough around the edges (albeit potentially useful as a developer tool).  For instance it seems to run into issues when buffers are deleted or `logms` is called in unanticipated ways like `(funcall #'logms "testing")`.  Personally I think these limitations should be addressed before we merge.  But it's also worth bearing in mind that sometimes even useful Elisp examples don't always belong in MELPA.  I'm open to other opinions on this, too.

In the meantime here's a few comments I jotted down while going through this:

- In most cases simply loading a package shouldn't do things like call `advice-add` -- if you catch yourself doing this, a global minor mode might be a good idea.  It lets users disable the behavior with a familiar toggle
- Using a `post-command-hook` to defer clearing the hash table seems a bit hacky to me. Why couldn't this happen at the end of `logms` itself?
- Minor, but the dependency on `s` is a little heavy (both functions used are more or less aliases to builtins)

**Comment by jcs090218** [comment]
> I see. I think the package summary "see where the message came from" creates some of the confusion and should probably be revised. I'd say this package provides a logging function whose output contains clickable links to context.

Changed! Thanks for the advice.

>  For instance it seems to run into issues when buffers are deleted or logms is called in unanticipated ways like (funcall #'logms "testing"). 

Yes, unfortunately, there is no way to track the deleted function since user can easily evaluate `logms` in `*scratch*` then delete it afterward. 

I did think about calling `logms` using funcall yet I don't think this can be useful in any circumstance. For instance, I use `funcall` with in the debug like 

```el
(when debug-flag
  (funcall #'logms fmt args))  ; the context will always be on this line and it's not useful
```

hence, 

1. User shouldn't use `logms` inside `funcall` since it doesn't really help you debug anything
2. I have tested with `funcall` and it works (but not with `apply` though, it stucks at `find-function` and I don't know how to exit it at the moment 😕)

Let me know if there is other use cases. 😅 

> * In most cases simply loading a package shouldn't do things like call advice-add -- if you catch yourself doing this, a global minor mode might be a good idea. It lets users disable the behavior with a familiar toggle

Thanks for the catch! Added!

> * Using a post-command-hook to defer clearing the hash table seems a bit hacky to me. Why couldn't this happen at the end of logms itself?

No, the hash-table should be cleaned after each command. Hash-table tracks down a full `logms` activities throughout the command, so you can log the same message but still get the correct **point** to navigate to. For example, if you `eval-buffer` the following

```el
(logms "hello")  ; if you clean hash-table here, you lost the data
(logms "hello")  ; this `logms` will be pointed to the last line (wrong)
```

Even you check with variable `last-command` or `this-command`, still you cannot guarantee the user uses the different command between each command calls to `logms`. And I reckon `post-command` is a clean solution since it only happens when `logms` get called, and it immediately removes it afterward.

**Comment by riscy** [comment]
A couple small fyi's: if you run package-lint against the new summary you chose, you'll see it's too long for MELPA.  And now that you've introduced a minor mode, an on/off variable like `logms-show` becomes redundant.

> User shouldn't use logms inside funcall since it doesn't really help you debug anything

Yeah, I can see what you mean, but I think it's hard to make that claim with 100% certainty.  But also, all of this inconsistent behavior isn't documented and it's left to the user (me in this case) to puzzle out why.

> I have tested with funcall and it works (but not with apply though, it stucks at find-function and I don't know how to exit it at the moment 😕)

Maybe it's not working in my particular case.  Here's the forms I'm evaluating:
```elisp
(require 'logms)
(logms-mode)
(setq log-function #'logms)  ; #'message)

(defun mytest ()
  (interactive)
  (funcall log-function "testing"))

(defun mytest2 ()
  (interactive)
  (logms "testing"))
```
Here's what my `*Messages*` buffer looks like:
image

There are also situations to consider such as, what if `logms` gets called as a result of someone invoking `run-with-idle-timer` and that sort of thing.

**Comment by jcs090218** [comment]
> A couple small fyi's: if you run package-lint against the new summary you chose, you'll see it's too long for MELPA. And now that you've introduced a minor mode, an on/off variable like logms-show becomes redundant.

I have removed the flag and changed it to `logms-mode` instead. Now it should be cleaner without redundant code. 👍 

> Maybe it's not working in my particular case. Here's the forms I'm evaluating:

Thanks for the details! There was a miscalculation from the frame level (`logms--frame-level-at-point`); that's why it's complaining the source missing. I have also added the support for `funcall` and `apply`.

> There are also situations to consider such as, what if logms gets called as a result of someone invoking run-with-idle-timer and that sort of thing.

Bad news. I have thought about this issue, and I think this is impossible to achieve at the moment. Since the end backtrace frame returns function `timer-event-handler`, I cannot know which function is called. It can either be `run-with-timer` or `run-with-idle-timer`. The timer event also leads to another issue, like the execution is being delay so I cannot locate the file that the time `logms` is being executed (not even inside eval history). 

In conclusion,

##### ✔️  Working

```el
(funcall #'logms "hello")   ; good
(apply #'logms '("hello"))  ; good
```

##### ❗  Not Working

```el
(run-with-idle-timer 1 nil (lambda () (logms "hello")))  ; bad
(run-with-idle-timer 1 nil #'logms "hello")              ; bad
```

I have temporarily disabled the timer event for now by informing the user not to use it inside the timer event. 😓 

https://github.com/jcs-elpa/logms/blob/0c3fdbaf368981dc75d55b9e59c37a048f420dd5/logms.el#L266

**Comment by riscy** [comment]
> Bad news. I have thought about this issue, and I think this is impossible to achieve at the moment.

Okay -- I think the effort is appreciated here, but a lot of these rough edges are the kinds of things you want to go over yourself (and be aware of) before asking for a review.  It puts me in a tough spot as a reviewer because I have my own projects I like to work on, too.

**Comment by jcs090218** [comment]
I will make sure I have fully tested the package before asking for the review next time! And thanks again!
